gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+    and its children of class :param:`node_type`.
+
+    :param node: root of expression
+    :param node_type: class of node (e.g. Sum or Product)
+    :return: list of all terms
+    """
+    from collections import deque
+    terms = []  # collected terms
+    queue = deque([node])  # queue of children nodes to process
+    while queue:
+        child = queue.popleft()
+        if isinstance(child, node_type):
+            queue.extendleft(reversed(child.children))
+        else:
+            terms.append(child)
+    return tuple(terms)
+
+
+def flatten_sum(node, argument_indices):
+    """
+    factorise :param:`node` into sum of products, group factors of each product
+    based on its dependency on index:
+    1) nothing (key = 0)
+    2) i (key = 1)
+    3) (i)j (key = j)
+    4) (i)k (key = k)
+    ...
+    5) (i)jk (key = 2)
+    :param node: root of expression
+    :param index: tuple of argument (linear) indices
+    :return: dictionary to list of factors
+    """
+    monos = collect_terms(node, Sum)
+    result = []
+    arg_ind_set = set(argument_indices)
+    for mono in monos:
+        d = OrderedDict()
+        for i in [0, 1, 2] + list(argument_indices):
+            d[i] = list()
+        for factor in collect_terms(mono, Product):
+            fi = factor.free_indices
+            if not fi:
+                d[0].append(factor)
+            else:
+                ind_set = set(fi) & arg_ind_set
+                if len(ind_set) > 1:
+                    # Aijk
+                    d[2].append(factor)
+                elif len(ind_set) == 0:
+                    # Ai
+                    d[1].append(factor)
+                else:
+                    # Aij
+                    d[ind_set.pop()].append(factor)
+            # elif j in fi and k in fi:
+            #     d[2].append(factor)
+            # elif j in fi:
+            #     d[j].append(factor)
+            # elif k in fi:
+            #     d[k].append(factor)
+            # else:
+            #     d[1].append(factor)
+        for i in d:
+            d[i] = tuple(d[i])
+        result.append(d)
+    return tuple(result)
+
+
+def sumproduct_2_node(factor_lists):
+    """
+    generate gem node from list of factors
+    use recursion so that each term is not too long
+    i.e. (a+b) + (c+d) instead of (((a+b)+c)+d
+    :param factor_lists: list of factors representing sums of products
+    :return: gem node
+    """
+    if len(factor_lists) < 3:
+        sums = []
+        for fl in factor_lists:
+            sums.append(reduce(Product, fl, one))
+        return reduce(Sum, sums, Zero())
+    else:
+        mid = int(len(factor_lists) / 2)
+        return Sum(sumproduct_2_node(factor_lists[:mid]), sumproduct_2_node(factor_lists[mid:]))
+
+
+def factorise(factor_lists, quad_ind, arg_ind_flat):
+    """
+    recursively pick the most common factor
+    maybe can Memoize this one
+    :param factors: list of list of factors, representing sum of products
```

Comment:
Docstring doesn't agree with what arguments are called.  Also, name the other two something like `quadrature_indices` and `argument_indices`? assuming that's what they are?